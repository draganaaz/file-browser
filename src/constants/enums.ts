export enum FILE_TYPE {
  FOLDER = 'folder',
  FILE = 'file',
}
