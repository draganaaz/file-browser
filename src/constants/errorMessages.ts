export const ERROR_MESSAGES = {
  INVALID_NAME_OR_EXTENSION: 'Invalid name or extension.',
  ITEM_EXISTS:
    'This item already exists at this location. Please choose a different name.',
  FILE_NOT_FOUND: 'File/Folder not found.',
};
