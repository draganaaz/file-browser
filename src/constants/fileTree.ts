export enum FILE_TYPE {
  FOLDER = 'folder',
  FILE = 'file',
}

export const VALID_EXTENSIONS = {
  TEXT: 'txt',
  JSON: 'json',
  IMAGE: 'png',
};

export const MENU_ITEMS = [
  { label: 'New Folder', type: FILE_TYPE.FOLDER },
  {
    label: 'New Text Document',
    type: FILE_TYPE.FILE,
    extension: VALID_EXTENSIONS.TEXT,
  },
  {
    label: 'New JSON Document',
    type: FILE_TYPE.FILE,
    extension: VALID_EXTENSIONS.JSON,
  },
  {
    label: 'New Image',
    type: FILE_TYPE.FILE,
    extension: VALID_EXTENSIONS.IMAGE,
  },
];
