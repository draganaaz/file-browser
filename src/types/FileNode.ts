import { FILE_TYPE } from '../constants/enums';

export interface FileNode {
  id: string;
  name: string;
  type: FILE_TYPE;
  children?: FileNode[];
}
