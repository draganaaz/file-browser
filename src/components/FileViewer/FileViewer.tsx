import React, { useEffect, useState, useCallback } from 'react';
import ReactQuill from 'react-quill';
import { FileNode } from '../../types/FileNode';
import { FILE_TYPE } from '../../constants/fileTree';
import 'react-quill/dist/quill.snow.css';
import { useFileTree } from '../../contexts/FileTreeContext';
import { updateFileContent } from '../../utils/treeService';

const FileViewer: React.FC = () => {
  const [content, setContent] = useState<string>('');

  const { selectedNode, setFileTree } = useFileTree();

  const handleUpdateContent = (id: string, updatedContent: string) => {
    setFileTree((prevTree) => updateFileContent(prevTree, id, updatedContent));
  };

  // Load content when selectedNode changes
  useEffect(() => {
    if (selectedNode?.type === FILE_TYPE.FILE && selectedNode.fileContent) {
      setContent(selectedNode.fileContent);
    }
  }, [selectedNode]);

  // Debounced content change handler
  const handleContentChange = useCallback((value: string) => {
    setContent(value);
  }, []);

  const handleSave = () => {
    if (selectedNode) {
      handleUpdateContent(selectedNode.id, content);
    }
  };

  const handleDiscard = () => {
    if (selectedNode?.type === FILE_TYPE.FILE) {
      setContent(selectedNode.fileContent || '');
    }
  };

  const renderFolderContent = (node: FileNode) => (
    <>
      <h2 className="text-lg font-bold mb-2" data-testid="file-name">
        {node.name}
      </h2>
      {node.children && node.children.length > 0 ? (
        <ul className="list-disc ml-6">
          {node.children.map((child) => (
            <li key={child.id} className="mb-1" data-testid="folder-item">
              {child.type === FILE_TYPE.FOLDER ? 'üìÅ' : 'üìÑ'} {child.name}
            </li>
          ))}
        </ul>
      ) : (
        <p className="text-gray-500">Folder is empty.</p>
      )}
    </>
  );

  const renderFileContent = (node: FileNode) => (
    <>
      <h2 className="text-lg font-bold mb-4" data-testid="file-name">
        {node.name}
      </h2>
      {node.name.endsWith('.png') ? (
        <img
          src={node.fileContent}
          alt={node.name}
          className="max-w-full h-auto border rounded"
          data-testid="image-file"
        />
      ) : (
        <>
          <ReactQuill
            value={content}
            onChange={handleContentChange}
            className="mb-6"
            theme="snow"
            data-testid="quill-editor"
          />
          <div className="flex justify-end gap-2 mt-4">
            <button
              onClick={handleSave}
              className="bg-blue-500 text-white px-4 py-2 rounded hover:bg-blue-600"
            >
              Save
            </button>
            <button
              onClick={handleDiscard}
              className="bg-red-500 text-white px-4 py-2 rounded hover:bg-red-600"
            >
              Discard
            </button>
          </div>
        </>
      )}
    </>
  );

  const renderContent = () => {
    if (!selectedNode) {
      return (
        <p data-testid="no-file-selected">
          Select a file or folder to view its contents.
        </p>
      );
    }
    return selectedNode.type === FILE_TYPE.FOLDER
      ? renderFolderContent(selectedNode)
      : renderFileContent(selectedNode);
  };

  return (
    <div className="p-4">
      <h2 className="text-lg font-semibold mb-2">File Viewer</h2>
      {renderContent()}
    </div>
  );
};

export default FileViewer;
