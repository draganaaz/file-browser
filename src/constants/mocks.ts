import { FileNode } from '../types/FileNode';
import { FILE_TYPE } from './fileTree';

export const initialData: FileNode[] = [
  {
    id: '1',
    name: 'public',
    type: FILE_TYPE.FOLDER,
    children: [
      {
        id: '1-1',
        name: 'exus-logo.png',
        type: FILE_TYPE.FILE,
        fileContent: '/exus.png',
      },
    ],
  },
  {
    id: '2',
    name: 'server',
    type: FILE_TYPE.FOLDER,
    children: [
      {
        id: '2-1',
        name: 'config.json',
        type: FILE_TYPE.FILE,
        fileContent: '{"key": "value"}',
      },
      {
        id: '2-2',
        name: 'notes.txt',
        type: FILE_TYPE.FILE,
        fileContent: 'This is a sample text file.',
      },
    ],
  },
  {
    id: '3',
    name: 'src',
    type: FILE_TYPE.FOLDER,
    children: [],
  },
];
