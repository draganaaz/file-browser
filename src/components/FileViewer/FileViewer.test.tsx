import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import FileViewer from './FileViewer'; // Adjust the import path if necessary
import { FILE_TYPE } from '../../constants/fileTree';
import { FileNode } from '../../types/FileNode';

const mockSetFileTree = jest.fn();
let mockSelectedNode: FileNode | null;

jest.mock('../../contexts/FileTreeContext', () => ({
  useFileTree: () => ({
    selectedNode: mockSelectedNode,
    setFileTree: mockSetFileTree,
  }),
}));

jest.mock('react-quill', () => {
  return function MockReactQuill({
    value,
    onChange,
  }: {
    value: string;
    onChange: (value: string) => void;
  }) {
    return (
      <textarea
        data-testid="quill-editor"
        value={value}
        onChange={(e) => onChange(e.target.value)}
      />
    );
  };
});

describe('FileViewer Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    mockSelectedNode = null;
  });

  it('renders no file selected message', () => {
    render(<FileViewer />);

    expect(screen.getByTestId('no-file-selected')).toBeInTheDocument();
  });

  it('renders folder contents', () => {
    mockSelectedNode = {
      id: '1',
      name: 'public',
      type: FILE_TYPE.FOLDER,
      children: [
        {
          id: '1-1',
          name: 'exus-logo.png',
          type: FILE_TYPE.FILE,
          fileContent: '/exus.png',
        },
      ],
    };
    render(<FileViewer />);

    expect(screen.getByText(mockSelectedNode.name)).toBeInTheDocument();
    expect(screen.getByTestId('folder-item')).toBeInTheDocument();
  });

  it('renders file contents', () => {
    mockSelectedNode = {
      id: '2-2',
      name: 'notes.txt',
      type: FILE_TYPE.FILE,
      fileContent: 'This is a sample text file.',
    };
    render(<FileViewer />);

    expect(screen.getByTestId('file-name')).toHaveTextContent(
      mockSelectedNode.name
    );
    expect(screen.getByTestId('quill-editor')).toHaveValue(
      mockSelectedNode.fileContent
    );
  });

  it('updates file content on save', () => {
    mockSelectedNode = {
      id: '2-2',
      name: 'notes.txt',
      type: FILE_TYPE.FILE,
      fileContent: 'This is a sample text file.',
    };
    render(<FileViewer />);

    const newContent = 'Updated content';
    const quillEditor = screen.getByTestId('quill-editor');
    fireEvent.change(quillEditor, { target: { value: newContent } });

    fireEvent.click(screen.getByText('Save'));
    expect(mockSetFileTree).toHaveBeenCalledWith(expect.any(Function));
  });

  it('reverts file content on discard', () => {
    mockSelectedNode = {
      id: '2-2',
      name: 'notes.txt',
      type: FILE_TYPE.FILE,
      fileContent: 'This is a sample text file.',
    };
    render(<FileViewer />);

    const quillEditor = screen.getByTestId('quill-editor');
    fireEvent.change(quillEditor, { target: { value: 'Temporary change' } });

    fireEvent.click(screen.getByText('Discard'));
    expect(quillEditor).toHaveValue(mockSelectedNode.fileContent);
  });

  it('renders image file', () => {
    mockSelectedNode = {
      id: '1-1',
      name: 'exus-logo.png',
      type: FILE_TYPE.FILE,
      fileContent: '/exus.png',
    };
    render(<FileViewer />);

    const image = screen.getByTestId('image-file');
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('src', mockSelectedNode.fileContent);
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<FileViewer />);
    expect(asFragment()).toMatchSnapshot();
  });
});
