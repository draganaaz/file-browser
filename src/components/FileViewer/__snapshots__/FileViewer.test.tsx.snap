// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileViewer Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="p-4"
  >
    <h2
      aria-live="polite"
      class="text-lg font-semibold mb-2"
    >
      File Viewer
    </h2>
    <p
      data-testid="no-file-selected"
      role="alert"
    >
      Select a file or folder to view its contents.
    </p>
  </div>
</DocumentFragment>
`;
